lib/build.ts

```diff
@@ -327,33 +318,26 @@ async function checkAllowDockerPlatformHandling(
 		// All images specify platform, let Docker receive `--platform`
 		return true;
 	} else if (imagesWithPlatformSupport.length > 0) {
-		// Mixed images
-		console.log(
-			'Warning: Dockerfile contains a mixes images that require a platform selection with ones that do not support it.\n' +
-				'The following images do not support platform selection:\n-' +
-				imagesWithoutPlatformSupport.join('\n-') +
-				'\n\n' +
-				'The following images require platform selection:\n-' +
-				imagesWithPlatformSupport.join('\n-') +
-				'\n\n' +
-				'To avoid a possible build error, the CLI has disabled platform selection. ' +
-				'As a result, the architecture of the machine where the Docker Engine ' +
-				'is running will be used to select the platform when pulling upstream images ' +
-				'which may result in "exec format error" at runtime. To avoid this warning, ' +
-				'update or replace the images that do not support platform selection.',
-		);
-	} else {
-		// Only images without platform support
-		console.log(
-			'Warning: Dockerfile contains a images do not support platform selection.\n-' +
-				imagesWithoutPlatformSupport.join('\n-') +
-				'\n\n' +
-				'To avoid a possible build error, the CLI has disabled platform selection. ' +
-				'As a result, the architecture of the machine where the Docker Engine ' +
-				'is running will be used to select the platform when pulling upstream images ' +
-				'which may result in "exec format error" at runtime. To avoid this warning, ' +
-				'update or replace the images that do not support platform selection.',
-		);
+		// Here we know that the service references at least 2 images
+		// (at least 1 with platform support and at least 1 without platform
+		// support). Therefore the service is not an "external image service"
+		// (`task.imageName` must be falsy), and it must have a multi-stage
+		// Dockerfile that references at least 2 images.
+		warn(`\
+Service '${task.serviceName}':
+  Multi-stage Dockerfile found with a mix of base images that require
+  CPU architecture selection and base images that do not support it.
+  The following base images do not support CPU architecture selection:
+  - ${imagesWithoutPlatformSupport.join('\n  - ')}
+  The following base images require CPU architecture selection:
+  - ${imagesWithPlatformSupport.join('\n  - ')}
+  As a result, the CPU architecture of the machine where the Docker Engine
+  is running will be used by default to select base images that require
+  architecture selection. This may result in incorrect architecture selection
+  and "exec format error" at runtime. It is usually possible to override the
+  architecture in the FROM line with e.g. "FROM --platform=linux/arm/v7",
+  or by adding the sha256 digest of the image for a specific architecture
+  with e.g. "FROM debian@sha256:094f57...".`);
```

Comment:
For another PR, we could detect whether the user is already following this advice to avoid printing the warning, i.e. platform-ready images in the Dockerfile are already either selecting the image with a sha256 digest (as support agents have been advising users to do for years, as a workaround for lack of multiarch support), or already using `FROM --platform`. It can be confusing to be warned to do something (use `FROM --platform`) when you are already doing it.

Reply:
Post merge task:

- [ ] Add an issue to address this

Reply:
See: https://github.com/balena-io-modules/balena-multibuild/issues/97